starters/express-apollo-prisma/src/graphql/data-sources/technology-data-source.ts

```diff
@@ -0,0 +1,97 @@
+import { PrismaClient, Prisma, TechnologyEntity } from '@prisma/client';
+import { RedisClient } from '../../redis';
+
+type TechnologyEntityId = TechnologyEntity['id'];
+
+export class TechnologyDataSource {
+	constructor(
+		private prismaClient: PrismaClient,
+		private redisClient?: RedisClient,
+		private redisCacheTtlSeconds: number = 360
+	) {}
+
+	async getTechnologyById(id: TechnologyEntityId): Promise<TechnologyEntity | null> {
+		let entity = await this.getCached(id);
+		if (entity) {
+			return entity;
+		}
+		entity = await this.prismaClient.technologyEntity.findFirst({
+			where: {
+				id,
+			},
+		});
+		if (entity) {
+			await this.cache(entity);
+		}
+		return entity;
+	}
+
+	async getTechnologies(): Promise<TechnologyEntity[]> {
+		return await this.prismaClient.technologyEntity.findMany();
+	}
+
+	async createTechnology(data: Prisma.TechnologyEntityCreateInput): Promise<TechnologyEntity> {
+		const entity = await this.prismaClient.technologyEntity.create({
+			data,
+		});
+		this.cache(entity);
+		return entity;
+	}
+
+	async updateTechnology(
+		id: TechnologyEntityId,
+		updateTechnology: Prisma.TechnologyEntityUpdateInput
+	): Promise<TechnologyEntity> {
+		const entity = await this.prismaClient.technologyEntity.update({
+			where: {
+				id,
+			},
+			data: updateTechnology,
+		});
+		await this.cache(entity);
```

Comment:
What happens if Redis is unreachable at this point?

Reply:
Although not what I was going for, that's interesting as well. Do we really want to block the whole application if we don't have caching set up? Could it do with a warning instead?

What I'm really after is, what happens if Redis dies while the stack is already running. Or what happens if there's a network hiccup and our Node instance fails to connect for this request?

Reply:
Here's my concern about the code. The health endpoint will tell our monitoring whether our infrastructure is fully healthy. It does not alter how the production code behaves. Consider the following scenario:

 1. We start the stack, and everything is healthy
 2. After a while, Redis goes down. Our health endpoint reflects this
 3. In time, we try to get Redis back up
 4. Meanwhile, people will make requests which will call `updateTechnology` or `getTechnologyById`
 5. **if I'm not missing anything**, these requests will throw an error (500, most likely) because the Redis connection is dead.
 
So what I would like to know is whether we need to protect against this scenario. It would be a shame if our API goes down simply because we lost caching capabilities. 
Do you agree that we need a ticket for this?